nats.go

```diff
@@ -3613,6 +3623,51 @@ func (nc *Conn) unsubscribe(sub *Subscription, max int, drainMode bool) error {
 	return nil
 }
 
+func (s *Subscription) handleControlMessage(msg *Msg) (bool, error) {
+	isControl := len(msg.Data) == 0 && msg.Header.Get(statusHdr) == controlMsg
+	if !isControl {
+		return isControl, nil
+	}
+
+	// If it is a flow control message then have to ack.
+	if msg.Reply != "" {
+		err := msg.Respond(nil)
+		if err != nil {
+			return isControl, err
+		}
+	}
+	return isControl, nil
+}
+
+// processControlFlow checks whether it is a control message and
+// handles it automatically.
+func (s *Subscription) processControlFlow(t *time.Timer, mch <-chan *Msg) (*Msg, error) {
+	// We will peek at the channel and return the next message
+	// that is not a control message or a timeout error.
+	var msg *Msg
+	var ok bool
+	for {
+		select {
+		case msg, ok = <-mch:
```

Comment:
There is only one of these now since passing the channels from the timer and ctx, in case of either being `nil` then it would just block allowing the other one to receive.